ipv8/test/test_requestcache.py

```diff
@@ -59,14 +60,16 @@ async def test_shutdown(self):
         """
         Test if RequestCache does not allow new Caches after shutdown().
         """
-        num_tasks = len(all_tasks())
-        request_cache = RequestCache()  # This adds a task, don't use ``self.request_cache`` here!
-        request_cache.add(MockCache(request_cache))
-        self.assertEqual(len(all_tasks()), num_tasks + 2)
-        await request_cache.shutdown()
-        self.assertEqual(len(all_tasks()), num_tasks)
-        request_cache.add(MockCache(request_cache))
-        self.assertEqual(len(all_tasks()), num_tasks)
+        num_tasks = len(all_tasks())  # [Background tasks (depends on test runner) + RequestCache]
+        self.request_cache.add(MockCache(self.request_cache))
+        self.assertEqual(len(all_tasks()), num_tasks + 1)  # [Background + RequestCache + Cache]
+        await self.request_cache.shutdown()
+        self.assertEqual(len(all_tasks()), num_tasks - 1)  # [Background]
+        with self.assertLogs(level=logging.WARNING) as log:
+            self.request_cache.add(MockCache(self.request_cache))  # No tasks should have been added
+        self.assertEqual(len(all_tasks()), num_tasks - 1)  # [Background]
+        self.assertEqual(len(log.records), 1)
+        self.assertEqual(log.records[0].msg, 'Dropping %s due to shutdown!')
```

Comment:
Your change is asserting "logging output" on a unit that is asserting the "number of asyncio tasks". Therefore, your change should be in a separate unit test.

My secondary concern with your change is that you're checking the logging message itself. Whereas I see "logging a warning"---though this is not done in any other IPv8 unit test yet---as an improvement over the current tests, the exact log message is an implementation detail that should not be asserted. [In my opinion this StackOverflow thread nicely covers why and the differing opinions on this topic.](https://softwareengineering.stackexchange.com/q/384164)

```suggestion
        self.request_cache.add(MockCache(self.request_cache))  # No tasks should have been added
        self.assertEqual(len(all_tasks()), num_tasks - 1)  # [Background]
        
    async def test_add_post_shutdown_warning(self):
        """
        Test if RequestCache logs a warning when adding a cache after shutdown.
        """
        await self.request_cache.shutdown()
        with self.assertLogs(level=logging.WARNING) as log:
            self.request_cache.add(MockCache(self.request_cache))
        self.assertGreaterEqual(len(log.records), 1)
```

Reply:
> Your change is asserting "logging output" on a unit that is asserting the "number of asyncio tasks". Therefore, your change should be in a separate unit test.

I'm not sure I understand correctly what you mean by "unit" here, but the unit test mentioned was not for testing the "number of asyncio tasks". It is named "test_shutdown" and tests the behavior of the "request_cache.shutdown()" call. From the docstring: "Test if RequestCache does not allow new Caches after shutdown()". Nothing in it assumes that the only check belonging to this test is for the "number of asyncio tasks". As issuing a log warning is part of the behavior of "request_cache.shutdown()", it looks like the proper place to test it.

> the exact log message is an implementation detail that should not be asserted

This was the reason I was testing the message template and not the **exact** log message, as the exact log message includes unstable repr of a cache object.

> this StackOverflow thread nicely covers why

The most popular answer in the threads tells that "you should absolutely test your logging" and that "you should test that the action of logging was done", "not the log files concretely". Testing the log record and not the resulted formatted output of a physical log stream is exactly what is advised here. Also, it is brittle just to test that the log action was done without checking the message, as it may be some unrelated log action, like an unrelated deprecation warning.

Anyway, I think it is not as important to test log messages here, so I removed the check. Now, the code exactly corresponds to your last proposal.

Reply:
My thoughts on this are as follows:

The log message is definitely part of the current implementation. It is definitely helpful when debugging. However, a programmer should _never_ depend on the template of the log message when writing code. Therefore, it should not be in a unit test.

I think that there is a case to be made that receiving a warning (without testing its mesage or message template) is part of the promise to our users ("behavior" as you call it). I agree that it is hard (impossible?) to test whether a log message orginates from any particular code and I agree with your resolution.